In [1]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.model_selection import train_test_split

In [2]:
data = pd.read_csv('train_data_cleaned_with_features.csv')
X = data.drop(columns = 'price')
y = data['price']

X_train, X_val, y_train, y_val = train_test_split(X,y, random_state = 1, test_size = 0.2)
X_train.head()

bed  bath  area  bintaro  serpong  bsd  ciputat  sektor  pamulang  \
18941  3.0   3.0   105        1        0    0        0       1         0   
6901   2.0   2.0    63        1        1    0        0       0         0   
9914   2.0   2.0    68        0        0    0        1       0         0   
9316   5.0   5.0   500        1        0    0        0       0         0   
9197   4.0   3.0   219        1        0    0        0       1         0   

       pondok  ...  SEKTOR  BSD  JAYA  SERPONG  LANTAI  DEKAT  MURAH  CITY  \
18941       1  ...       1    0     0        0       0      0      0     0   
6901        0  ...       0    0     0        0       0      0      0     0   
9914        0  ...       0    0     0        0       0      0      0     0   
9316        0  ...       0    0     0        0       0      0      0     0   
9197        0  ...       1    0     0        0       0      0      0     0   

       PAMULANG  CIPUTAT  
18941         0        0  
6901          0        0  
9914          0        0  
9316          0        0  
9197          0        0  

[5 rows x 26 columns]

In [3]:
forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(X_train, y_train)
predictions = forest_model.predict(X_val)
print("validation MAE: ", round(mean_absolute_error(y_val, predictions)))

validation MAE:  741573574


In [4]:
test_data = pd.read_csv('test_data_cleaned_with_features.csv')
X_test = test_data.drop(columns = 'price')
y_test = test_data['price']
preds = forest_model.predict(X_test)
print("test MAE: ", round(mean_absolute_error(y_test, preds)))

test MAE:  426821771


In [5]:
MAPE = round(100*mean_absolute_percentage_error(y_test,preds))
print("test Mean Absolute Prediction Error:", MAPE, "%")

test Mean Absolute Prediction Error: 26 %


I'd also like to see how the model performs without the dummy variables.

In [7]:
X = data[['bed','bath','area']]
y = data['price']

X_train, X_val, y_train, y_val = train_test_split(X,y, random_state = 1, test_size = 0.2)
X_train.head()

bed  bath  area
18941  3.0   3.0   105
6901   2.0   2.0    63
9914   2.0   2.0    68
9316   5.0   5.0   500
9197   4.0   3.0   219

In [8]:
new_forest_model = RandomForestRegressor(random_state=1)
new_forest_model.fit(X_train, y_train)
predictions = new_forest_model.predict(X_val)

print("validation MAE: ", round(mean_absolute_error(y_val, predictions)))

validation MAE:  814923853


In [9]:
X_test = test_data[['bed','bath','area']]
y_test = test_data['price']
preds = new_forest_model.predict(X_test)

print("test MAE: ", round(mean_absolute_error(y_test, preds)))

test MAE:  682260192


Which means our dummy variables brought down the MAE by (682260192 - 426821771)/682260192 = 37%